src/zarr/core/group.py

```diff
@@ -2774,11 +2860,360 @@ def array(
+
+            zarr_format = sample.zarr_format
+            # TODO: this type hint is so long
+            func: (
+                Callable[[Store, str], Coroutine[Any, Any, GroupMetadata | ArrayV3Metadata]]
+                | Callable[[Store, str], Coroutine[Any, Any, GroupMetadata | ArrayV2Metadata]]
+            )
+            if zarr_format == 3:
+                func = _read_metadata_v3
+            elif zarr_format == 2:
+                func = _read_metadata_v2
+            else:  # pragma: no cover
+                raise ValueError(f"Invalid zarr_format: {zarr_format}")  # pragma: no cover
+
+            coros = (func(store=store, path=_join_paths([path, key])) for key in nodes_parsed)
+            extant_node_query = dict(
+                zip(
+                    nodes_parsed.keys(),
+                    await asyncio.gather(*coros, return_exceptions=True),
+                    strict=False,
+                )
+            )
+
+            for key, value in extant_node_query.items():
+                if isinstance(value, BaseException):
+                    if isinstance(value, FileNotFoundError):
+                        # ignore FileNotFoundError, because they represent nodes we can safely create
+                        pass
+                    else:
+                        # Any other exception is a real error
+                        raise value
+                else:
+                    # this is a node that already exists, but a node with this name was specified in
+                    # nodes_parsed.
+                    # Two cases produce exceptions:
+                    # 1. The node is a group, and a node with this name was explicitly defined in
+                    # nodes
+                    # 2. The node is an array.
+                    # The third case is when this extant node is a group, but its name was not
+                    # explicitly defined in nodes. This means it was added as an implicit group by
+                    # _parse_hierarchy_dict, and we can remove the reference to this node from
+                    # nodes_parsed. We don't need to create this node.
+
+                    if isinstance(value, GroupMetadata):
+                        if key not in implicit_group_names:
+                            raise ContainsGroupError(store, key)
+                        else:
+                            # as there is already a group with this name, we should not create a new one
+                            redundant_implicit_groups.append(key)
+                    elif isinstance(value, ArrayV2Metadata | ArrayV3Metadata):
+                        raise ContainsArrayError(store, key)
+
+            nodes_parsed = {
+                k: v for k, v in nodes_parsed.items() if k not in redundant_implicit_groups
+            }
+
+    async for node in create_nodes(store=store, path=path, nodes=nodes_parsed):
+        yield node
+
+
+async def create_nodes(
```

Comment:
advocating to make this a private function and not export it to `zarr.api`.

Reply:
What's the position on using private zarr functions from within xarray?

Reply:
it's now private